test/service_test.cc

```diff
@@ -0,0 +1,458 @@
+};
+
+// This tests on `registeruser` in Service Server
+TEST_F(DISABLED_ServiceTestServer, RegisterUser) {
+  // Try to register existed usernames
+  for(size_t i = 0; i < kNumOfUsersPreset; ++i) {
+    // ServiceClient::ReturnCodes
+    auto ret = service_client_.SendRegisterUserRequest(user_list_[i]);
+    // This should fail since the username specified has already been registered in the `SetUp` process above
+    EXPECT_EQ(ServiceClient::USER_EXISTS, ret) << "This should fail since the username specified has been registered.";
+  }
+
+  // Try to register non-existed usernames
+  for(size_t i = kNumOfUsersPreset; i < kNumOfUsersTotal; ++i) {
+    // ServiceClient::ReturnCodes
+    auto ret = service_client_.SendRegisterUserRequest(user_list_[i]);
+    // This should succeed since the username specified is not registered
+    EXPECT_EQ(ServiceClient::OK, ret) << "This registration should succeed.";
+  }
+
+  // Try to register all usernames again
+  for(size_t i = 0; i < kNumOfUsersTotal; ++i) {
+    // ServiceClient::ReturnCodes
+    auto ret = service_client_.SendRegisterUserRequest(user_list_[i]);
+    // This should fail since all these usernames are already registered
+    EXPECT_EQ(ServiceClient::USER_EXISTS, ret) << "This should fail since the username specified has been registered.";
+  }
+}
+
+// This tests on `chirp` in Service Server
+TEST_F(DISABLED_ServiceTestServer, Chirp) {
+  // Every user posts a chirp
+  uint64_t last_id = 0;
+  for (size_t i = 0; i < kNumOfUsersTotal; ++i) {
+    struct ServiceClient::Chirp chirp;
+    // ServiceClient::ReturnCodes
+    auto ret = service_client_.SendChirpRequest(user_list_[i], kShortText, last_id, &chirp);
+    EXPECT_EQ(ServiceClient::OK, ret);
+    // The three lines below check if the data in a chirp data structure corresponds to
+    // the data provided above
+    EXPECT_EQ(user_list_[i], chirp.username);
+    EXPECT_EQ(kShortText, chirp.text);
+    EXPECT_EQ(last_id, chirp.parent_id);
+    last_id = chirp.id;
+  }
+}
+
+// This tests on `follow` in Service Server
+TEST_F(DISABLED_ServiceTestServer, Follow) {
+  // Every user follows its next user
+  for(size_t i = 0; i < kNumOfUsersTotal; ++i) {
+    // ServiceClient::ReturnCodes
+    auto ret = service_client_.SendFollowRequest(user_list_[i], user_list_[(i + 1) % kNumOfUsersTotal]);
+    // This should be successful
+    EXPECT_EQ(ServiceClient::OK, ret);
+    // This should be unsuccessful since it tries to follow a non-existed user
+    ret = service_client_.SendFollowRequest(user_list_[i], "non-existed");
+    EXPECT_EQ(ServiceClient::FOLLOWEE_NOT_FOUND, ret);
+  }
+}
+
+// This tests on `read` in Service Server
+TEST_F(DISABLED_ServiceTestServer, Read) {
+  // The container of chirp ids
+  std::vector<uint64_t> corrected_chirps;
+
+  for(size_t i = 0; i < kNumOfUsersTotal; ++i) {
+    corrected_chirps.clear();
+    ServiceClient::Chirp chirp;
+
+    // In the following three layers, I try to build a nested chirps like this
```

Comment:
I'm not sure how to read this. Top down or left to right?

Reply:
I want to express that ```#02```, ```#03```, and ```#04``` reply to ```#01```, and ```#05```, ```#06```, and ```#07``` reply to ```#02```.